core/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -0,0 +1,156 @@
+     */
+    private final LinkedTransferQueue<Tuple<Translog.Location, Consumer<Boolean>>> refreshListeners = new LinkedTransferQueue<>();
+
+    private final IntSupplier getMaxRefreshListeners;
+    private final Runnable forceRefresh;
+    private final Consumer<Runnable> fireListener;
+
+    /**
+     * The estimated size of refreshListenersEstimatedSize used for triggering refresh when the size gets over
+     * {@link IndexSettings#MAX_REFRESH_LISTENERS_PER_SHARD}. No effort is made to correct for threading issues in the size calculation
+     * beyond it being volatile.
+     */
+    private volatile int refreshListenersEstimatedSize;
+    /**
+     * The translog location that was last made visible by a refresh.
+     */
+    private volatile Translog.Location lastRefreshedLocation;
+
+    public RefreshListeners(IntSupplier getMaxRefreshListeners, Runnable forceRefresh, Consumer<Runnable> fireListener) {
+        this.getMaxRefreshListeners = getMaxRefreshListeners;
+        this.forceRefresh = forceRefresh;
+        this.fireListener = fireListener;
+    }
+
+    /**
+     * Add a listener for refreshes.
+     *
+     * @param location the location to listen for
+     * @param listener for the refresh. Called with true if registering the listener ran it out of slots and forced a refresh. Called with
+     *        false otherwise.
+     */
+    public void add(Translog.Location location, Consumer<Boolean> listener) {
+        requireNonNull(listener, "listener cannot be null");
+        Translog.Location listenerLocation = requireNonNull(location, "location cannot be null");
+
+        Translog.Location lastRefresh = lastRefreshedLocation;
+        if (lastRefresh != null && lastRefresh.compareTo(listenerLocation) >= 0) {
+            // Location already visible, just call the listener
+            listener.accept(false);
+            return;
+        }
+        if (refreshListenersEstimatedSize < getMaxRefreshListeners.getAsInt()) {
+            // We have a free slot so register the listener
+            refreshListeners.add(new Tuple<>(location, listener));
+            refreshListenersEstimatedSize++;
+            return;
+        }
+        /*
+         * No free slot so force a refresh and call the listener in this thread. Do so outside of the synchronized block so any other
+         * attempts to add a listener can continue.
+         */
+        forceRefresh.run();
+        listener.accept(true);
+    }
+
+    /**
+     * Start listening to an engine.
+     */
+    public void listenTo(Engine engine) {
+        engine.registerSearchRefreshListener(new RefreshListenerCallingRefreshListener(engine));
+    }
+
+    /**
+     * Listens to Lucene's {@linkplain ReferenceManager.RefreshListener} and fires off listeners added by
+     * {@linkplain IndexShard#addRefreshListener(Translog.Location, Consumer)}.
+     */
+    private class RefreshListenerCallingRefreshListener implements ReferenceManager.RefreshListener {
```

Comment:
I don't like having a the class implementing an interface that is about it's implementation rather than its API. Would it be ok if I made the listener final?